File: Day286/findModeInBST.java
import java.util.ArrayList;
import java.util.List;

import javax.swing.tree.TreeNode;

class Solution {
    public int fre;
    public int temp = 0;
    public int maxCount = 0;
    public List<Integer> list = new ArrayList<>();

    public int[] findMode(TreeNode root) {
        inorder(root);
        int[] answer = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        return answer;
    }

    public void inorder(TreeNode node) {
        if (node == null) return;

        inorder(node.left);

        temp = (node.val == fre) ? temp + 1 : 1;
        if (temp == maxCount) {
            list.add(node.val);
        } else if (temp > maxCount) {
            maxCount = temp;
            list.clear();
            list.add(node.val);
        }
        fre = node.val;

        inorder(node.right);
    }
}

File: Day286/findModeInBST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func inorder(root *TreeNode, m map[int]int) {
    if root == nil {
        return
    }
    inorder(root.Left, m)
    m[root.Val]++
    inorder(root.Right, m)
}
func findMode(root *TreeNode) []int {
    m := make(map[int]int)
    inorder(root, m)
    max := -1
    for _,v := range m {
        max = int(math.Max(float64(max), float64(v)))
    }
    answer := make([]int, 0)
    for k,v := range m {
        if v == max {
            answer = append(answer, k)
        }
    }
    return answer
}